src/Database/Query.php

```diff
@@ -889,6 +889,46 @@ public function where($conditions = null, $types = [], $overwrite = false)
         return $this;
     }
 
+    /**
+     * Adds an IN condition or set of conditions to be used in the WHERE clause for this
+     * query.
+     *
+     * This method does allow empty inputs in contrast to where().
+     * Be careful about using it without proper sanity checks.
+     *
+     * @param string $field Field
+     * @param array $values Array of values
+     * @return $this
+     */
+    public function whereIn($field, array $values)
+    {
+        if (!$values) {
+            return $this->where('1=0');
+        }
+
+        return $this->where([$field . ' IN' => $values]);
+    }
+
+    /**
+     * Adds a NOT IN condition or set of conditions to be used in the WHERE clause for this
+     * query.
+     *
+     * This method does allow empty inputs in contrast to where().
+     * Be careful about using it without proper sanity checks.
+     *
+     * @param string $field Field
+     * @param array $values Array of values
+     * @return $this
+     */
+    public function whereNotIn($field, array $values)
+    {
+        if (!$values) {
+            return $this->where('1=1');
```

Comment:
I think it should be based on the previous discussion (#7092).

In the past, this `'NOT IN' => []` returned `1 != 1` and it was an intended behavior. Since it was not my expectation, I wanted to get exceptions in such cases. I thought it was better than `1 != 1`.

But if you think `'NOT IN' => []` should return `1 = 1`, I think we should bring up #7092 again and we shouldn't add a wrapper to avoid discussion. Since `1 = 1` is my first expectation, I wouldn't be opposed to return it. `IS NOT NULL` may be better though.

Reply:
Generating a `1=1` condition on empty data is a dangerous default. Passing an empty list could be a mistake that we should protect against to be consistent with our previous decisions as @chinpei215 pointed out.

Reply:
OK, I hear you @markstory 
But my experience tells me exactly the opposite:
The behavior is intended like this in pretty much every case, and the result is usually the "blowup" of the application in a later state where the condition is not met anymore then for some reason.
Sure, less problematic as applying a query to too much rows.

I adjusted the code to opt-in the intuitive behavior here for RAD, keeping the default consistent to the strict way it is in where().

I would much rather have this code:
```
$query = $this->find()
    ->where(['x' => y])
    ->whereIn('field', $values, true)
    ->orderDesc('x')
    ->all();
```
Then
```
if (!$values) {
    return [];
}
$query = $this->find()
    ->where(['x' => y])
    ->whereIn('field', $values)
    ->orderDesc('x')
    ->all();
```
Especially when you start nesting with OR etc, as the complexity of the "false" case because super error prone.

When the performance is not as relevant as readability/correctness of the code, I want to not have to add application logic on top everywhere.
Using the allowEmpty option gives a dev that deliberate choice now - nothing to protect him anymore then?

Reply:
Could we have `Query::ALLOW_EMPTY` (true) and `Query::DISALLOW_EMPTY` (false)?